x-pack/legacy/plugins/file_upload/public/util/file_parser.js

```diff
@@ -7,39 +7,120 @@
 import _ from 'lodash';
 import { geoJsonCleanAndValidate } from './geo_json_clean_and_validate';
 import { i18n } from '@kbn/i18n';
+const oboe = require('oboe');
 
-export async function readFile(file) {
-  const readPromise = new Promise((resolve, reject) => {
-    if (!file) {
-      reject(new Error(i18n.translate(
-        'xpack.fileUpload.fileParser.noFileProvided', {
-          defaultMessage: 'Error, no file provided',
-        })));
+// In local testing, performance improvements leveled off around this size
+export const FILE_BUFFER = 1024000;
+
+const readSlice = (fileReader, file, start, stop) => {
+  const blob = file.slice(start, stop);
+  fileReader.readAsBinaryString(blob);
+};
+
+const createOboeStreamAndPatterns = (cleanAndValidate, featureTracking) => {
+  const oboeStream = oboe();
+  oboeStream.node({
+    'features.*': feature => {
+      // TODO: Add handling and tracking for cleanAndValidate fails
+      featureTracking();
+      return cleanAndValidate(feature);
     }
-    const fr = new window.FileReader();
-    fr.onload = e => resolve(e.target.result);
-    fr.onerror = () => {
-      fr.abort();
+  });
+  return oboeStream;
+};
+
+let previousFileReader;
+export const fileHandler = (
+  file, chunkHandler, cleanAndValidate, getFileParseActive,
+  fileReader = new FileReader(), fileBuffer = FILE_BUFFER, oboeStream
+) => {
+
+  if (!file) {
+    return Promise.reject(
+      new Error(
+        i18n.translate('xpack.fileUpload.fileParser.noFileProvided', {
+          defaultMessage: 'Error, no file provided',
+        })
+      )
+    );
+  }
+
+
+  // Halt any previous file reading activity
+  if (previousFileReader) {
+    previousFileReader.abort();
+  }
+
+  let start;
+  let stop = fileBuffer;
+  previousFileReader = fileReader;
+
+  let updateProgress;
+  if (!oboeStream) {
```

Comment:
I don't understand. Is oboeStream ever passed in as a param?

Reply:
Yeah, this update was made in response to some previous feedback which resulted in pulling the oboeStream declaration out of the function args...but we still needed to check its existence for tests. Long story short, it wasn't a very clean fix and I've updated it to use an instance of a mockable `PatternReader` which looks a lot better than this weirdness